################################################
# Helper file containing functions for conducting
# methylation analysis on BedGraph files produced
# from the Bismark Alignment suite.
#
# Please contact Ryan Quan if you have any
# questions regarding this file at rcq2102@columbia.edu
################################################


## choose directory of bedGraph files 
## and read into R as BSseq object 
read_bismark <- function() {
    
    fileNames <- list.files(path="data", pattern = ".bedGraph", full.names=TRUE)
    sampleNames <- sapply(fileNames,
                          # gets ID in between underscores
                          function(X) str_match(X, "_(.*?)_")[2],
                          USE.NAMES = FALSE)
    
    BSseq_obj <- read.bismark(files = fileNames, sampleNames = sampleNames,
                              rmZeroCov = FALSE, verbose = TRUE)
    # ensures the format of chromosome numbers to be "chr#"
    chr_names <- levels(BSseq_obj@rowData@seqnames@values)
    for (i in 1:length(chr_names)) {
        if (!str_detect(chr_names[i], "chr")) {
            chr_names[i] <- paste("chr", chr_names[i], sep = "")
            print(chr_names[i])
        }
    }
    return(BSseq_obj)
}

## assigns phenotypic data (e.g. tumor/normal) and
## excludes methylation loci below a minimum threshold by type
subset_by_type <- function(BSmooth_obj, min_cov = 300) {
    
    fileNames <- list.files(path="data", pattern = ".bedGraph")
    sampleNames <- sapply(fileNames,
                          # gets ID in between underscores
                          function(X) str_match(X, "_(.*?)_")[2],
                          USE.NAMES = FALSE)   
    print(sampleNames)
    
    ## creating phenotypic data
    cat("Creating phenotypic data...")
    
    # pulls text characters from sampleName
    type <- tolower(str_extract(sampleNames, "[A-z]+"))
    # get user input to classify sample type
    type[type == "n"] <- "normal"
    type[type == "t"] <- "tumor"
    type[type == "ca"] <- "case"
    type[type == "co"] <- "control"
    
    pData <- as.data.frame(sampleNames, row.names = NULL)
    pData <- cbind(pData, type)
    pData(BSmooth_obj) <- pData
    
    # assign color to samples (blue = normal/control, red = tumor/case)
    pData <- pData(BSmooth_obj)
    pData$col <- ifelse(type %in% c("control", "normal"), "blue", "red")
    pData(BSmooth_obj) <- pData
    
    ## subsetting by type
    cat(sprintf("Filtering methylation loci with less than %i reads...", min_cov))
    
    methCov <- getCoverage(BSmooth_obj)
    # filter out those with less than min_cov
    keepLoci <- which(rowSums(
        methCov[, BSmooth_obj$type %in% c("control", "normal")] >= min_cov) >= 1 &
            rowSums(methCov[, BSmooth_obj$type %in% c("case", "tumor")] >= min_cov) >= 1)
    BSmooth_obj <- BSmooth_obj[keepLoci,]
    
    return(BSmooth_obj)
}

get_tstat <- function(BSmooth_obj, est_var = "group2") {
    
    group1 <- which(pData(BSmooth_obj)$type %in% c("case", "tumor"))
    group2 <- which(pData(BSmooth_obj)$type %in% c("control","normal"))
    
    BStstat_obj <- BSmooth.tstat(BSmooth_obj, group2 = group2, # normal
                                 group1 = group1, # tumor
                                 estimate.var = est_var,
                                 local.correct = TRUE,
                                 verbose = TRUE)
    # removes missing t-stat values generated by BSmooth.tstat
    missing_values <- which(is.na(getStats(BStstat_obj)))
    BStstat_obj <- BStstat_obj[-missing_values,]
    
    return(BStstat_obj)
}

export_tstat_data <- function(BStstat_obj) {
    
    ## output marginal distribution of t-stat
    cat("Plotting marginal distribution of the t-statistic...\n")
    png("results/plots/tstat_marginalDistribution.png")
    plot(BStstat_obj) 
    title("Marginal Distribution of T-Statistic")
    dev.off()
    
    ## output table of t-stats for each methylation loci
    cat("Exporting table of t-statistics to results directory...\n")
    stats <- cbind(BStstat_obj@gr@ranges@start, getStats(BStstat_obj))
    write.csv(stats, "results/tables/tstat_ttestValues.csv")
}

export_dmr_data <- function(BSmooth_obj, BStstat_obj,
                            batch_num = c(1, 2),
                            FDR = 0.05, 
                            max_gap = 100, 
                            cg_num = 1,
                            mean_diff = 0.1,
                            min_cov = 300) {
    cat("Exporting table of DMRs to results directory...\n")
    cat(sprintf("FDR < %.2f...\n", FDR))
    cat(sprintf("Subsetting for DMRs with %i or more CpG sites and greater than %.2f 
                mean methylation difference.\n", cg_num, mean_diff))
    ## write table of DMRs
    dmrs0 <- dmrFinder(BStstat_obj, 
                       qcutoff = c(FDR, 1-FDR), 
                       maxGap = max_gap)
    write.csv(dmrs0, "results/tables/dmr_all.csv")
    
    ## write table of DMRs above specified mean methylation difference
    dmrs <- dmrs0[dmrs0$n >= cg_num & abs(dmrs0$meanDiff) >= mean_diff,]
    write.csv(dmrs, "results/tables/dmr_subset.csv")
    
    ## create DMR plot for each amplicon region
    
    amplicon_file <- list.files(pattern = ".csv")
    
    if (length(amplicon_file) == 0 | length(amplicon_file) >=2) {
        cat("Amplicon CSV file not detected in working directory. Generating plots for each DMR...\n")
        
        for (i in 1:nrow(dmrs)) {
            png(filename = paste("results/plots/dmr", i, ".png", sep = ""))
            suppressWarnings(plotRegion(BSmooth_obj,
                                        region = dmrs[i,],
                                        addPoints = TRUE,
                                        BSseqTstat = BStstat_obj,
                                        pointsMinCov = min_cov))
            dev.off()
        }
        
    } else if (length(amplicon_file) == 1) {
        
        amplicon <- read.csv(amplicon_file, stringsAsFactors = FALSE)
        amplicon_batch <- amplicon[amplicon$batch %in% batch_num,]
        
        cat("Generating DMR plots for the following amplicon regions:")
        print(amplicon_batch$amplicon)
        
        for (i in 1:nrow(amplicon_batch)) {
            png(filename = paste("results/plots/dmr_", amplicon_batch[i,1], ".png", sep = ""))
            suppressWarnings(plotRegion(BSmooth_obj, 
                                        region = amplicon_batch[i,c(3:5)],
                                        addPoints = TRUE,
                                        BSseqTstat = BStstat_obj,
                                        addRegions = dmrs,
                                        pointsMinCov = min_cov))
            title(xlab = amplicon_batch[i, 1])
            dev.off()
        }
    } 
}
